sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1528,8 +1528,8 @@ object EliminateSorts extends Rule[LogicalPlan] {
       applyLocally.lift(child).getOrElse(child)
     case s @ Sort(_, global, child) => s.copy(child = recursiveRemoveSort(child, global))
     case j @ Join(originLeft, originRight, _, cond, _) if cond.forall(_.deterministic) =>
-      j.copy(left = recursiveRemoveSort(originLeft, true),
-        right = recursiveRemoveSort(originRight, true))
+      j.copy(left = recursiveRemoveSort(originLeft, true, canEliminateThroughLocalLimit = true),
+        right = recursiveRemoveSort(originRight, true, canEliminateThroughLocalLimit = true))
     case g @ Aggregate(_, aggs, originChild) if isOrderIrrelevantAggs(aggs) =>
       g.copy(child = recursiveRemoveSort(originChild, true))
```

Comment:
I think aggregate can eliminate sort through local limit as well. The point is, local limit does not guarantee any semantic (the result depends on the data partitioning) and end-users won't have any expectations for it.